posts/2018-08-25-http4s-error-handling-mtl.md

```diff
@@ -0,0 +1,327 @@
+
+So let's get started by putting up an example of an http application with three different endpoints that interacts with a `UserAlgebra` that may or may not fail with some specific errors.
+
+If you are one of those who don't like to read and prefer to jump straight into the code please find it [here](https://gist.github.com/gvolpe/3fa32dd1b6abce2a5466efbf0eca9e94) :)
+
+### User Algebra
+
+We have a simple `UserAlgebra` that let us perform some actions such as finding and persisting users.
+
+```tut:book:silent
+case class User(username: String, age: Int)
+case class UserUpdateAge(age: Int)
+
+trait UserAlgebra[F[_]] {
+  def find(username: String): F[Option[User]]
+  def save(user: User): F[Unit]
+  def updateAge(username: String, age: Int): F[Unit]
+}
+```
+
+And also an ADT of the possible errors that may arise. I'll explain later in this post why it extends `Throwable`.
+
+```tut:book:silent
+sealed trait UserError extends Throwable
+case class UserAlreadyExists(username: String) extends UserError
+case class UserNotFound(username: String) extends UserError
+case class InvalidUserAge(age: Int) extends UserError
+```
+
+### User Interpreter
+
+And here we have a simple interpreter for our `UserAlgebra` for demonstration purposes so you can have an idea on how the logic would look like. In a real-life project an interpreter will more likely have some `Repository` as a dependency instead of using a `Ref`.
+
+```tut:book:silent
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+
+object UserInterpreter {
+
+  def create[F[_]](implicit F: Sync[F]): F[UserAlgebra[F]] =
+    Ref.of[F, Map[String, User]](Map.empty).map { state =>
+      new UserAlgebra[F] {
+        private def validateAge(age: Int): F[Unit] =
+          if (age <= 0) F.raiseError(InvalidUserAge(age)) else F.unit
+
+        override def find(username: String): F[Option[User]] =
+          state.get.map(_.get(username))
+
+        override def save(user: User): F[Unit] =
+          validateAge(user.age) *>
+            find(user.username).flatMap {
+              case Some(_) =>
+                F.raiseError(UserAlreadyExists(user.username))
+              case None =>
+                state.update(_.updated(user.username, user))
+            }
+
+        override def updateAge(username: String, age: Int): F[Unit] =
+          validateAge(age) *>
+            find(username).flatMap {
+              case Some(user) =>
+                state.update(_.updated(username, user.copy(age = age)))
+              case None =>
+                F.raiseError(UserNotFound(username))
+            }
+      }
+    }
+
+}
+```
+
+### Http Routes
+
+Right before we define the `UserRoutes` we need to write some generic Json codecs to avoid repeating ourselves:
+
+```tut:book:silent
+import io.circe._
+import io.circe.generic.auto._
+import org.http4s._
+import org.http4s.circe._
+
+object JsonCodecs {
```

Comment:
This exists as [CirceEntityCodecs](https://github.com/http4s/http4s/blob/v0.19.0-M1/circe/src/main/scala/org/http4s/circe/CirceEntityCodec.scala) now.

Reply:
Amazing! Didn't know about it :)